### PR TITLE
New config HOST_FOR_REDIRECTS to prevent host header injection

### PR DESCRIPTION
**Pull request by alanhamlett**
For #353.

Uses a new config option `HOST_FOR_REDIRECTS` on the current app.

### CODE REVIEW

**Comment by alanhamlett** [comment]
@Bastian-Kuhn can you take a look and check this PR?

**Comment by Bastian-Kuhn** [comment]
@alanhamlett of course, next week in the office i double check.

**Comment by Bastian-Kuhn** [comment]
Sorry for the delay, this change works for me. When the config option is set, the redirect can still be manipulated if for example the X-Forwaded-Host is injected. They only fix who works for me is the one i done, checking the request.url and throw an exception.

**Comment by davidism** [comment]
Why do you have a setup that allows injecting X-Forwarded headers? That's a security issue on your side.

File: flask_login/utils.py [hunk 1]
```diff
@@ -113,7 +113,10 @@ def login_url(login_view, next_url=None, next_field='next'):
     parsed_result = urlparse(base)
     md = url_decode(parsed_result.query)
     md[next_field] = make_next_param(base, next_url)
-    parsed_result = parsed_result._replace(query=url_encode(md, sort=True))
+    netloc = current_app.config.get('HOST_FOR_REDIRECTS') or \
+        parsed_result.netloc
+    parsed_result = parsed_result._replace(netloc=netloc,
+                                           query=url_encode(md, sort=True))
```
Comment (davidism) [inline flask_login/utils.py]: Isn't the issue with the `next` param? Why are we replacing the login url?

**Comment by Bastian-Kuhn** [comment]
The problem is that all our applications a reviewed by an HP Tool. This tool try to set the x-forwarded-host and tests if the application uses host header constructed hyperlinks, what flask_login does.

And this leads to an Issue with the classification High. That means that every application using flask_login will fail with this test. And for example one of our customers, a Worldwide known marked leader, every application has to pass this test in order to go (and stay) live. For me it's not a big deal to "fix" these locally, but i think this should be fixed out of the box.

**Comment by davidism** [comment]
OK, what I'm saying is that if anything setting X-Forwarded from the outside except your trusted server (Nginx) gets through to Flask, that's an issue with your server setup. Also, again, Flask-Login doesn't use X-Forwarded headers anywhere.

**Comment by davidism** [comment]
Since you seem to have a concrete test case in mind, would you please post a minimal example demonstrating the issue.

**Comment by Bastian-Kuhn** [comment]
Sure,
the App tries the following request
```
GET /auth/?has_age=True HTTP/1.1
Host: my-domain.de
x-forwarded-host: www.hpe.com
User-Agent: Mozilla/5.0 (Windows NT 6.3; WOW64; rv:53.0) Gecko/20100101 Firefox/53.0 Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8 Accept-Language: en-US,en;q=0.5
Referer: https://my-domain.de/?next=%2F
Cookie: session=XXX
```

And the servers (cause of flask_login) response is:
```
HTTP/1.1 302 FOUND
Server: nginx
Date: Tue, 30 May 2017 21:27:24 GMT
Content-Type: text/html; charset=utf-8
Content-Length: 209
Connection: close
Location: https://www.hpe.com/
Set-Cookie: session=XXX;
Secure; HttpOnly; Path=/
Strict-Transport-Security: max-age=63072000; includeSubdomains;
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
Cache-Control: private, max-age=0, s-maxage=0
```

That can reproduced in chrome with the extension requestly.
The x-forwarded-host: www.hpe.com is the problem here, and i don't see a good way to block this in nginx. I disagree with some of the Issues of this tool, but the part that flask login directly uses information based on host headers are right. On the other hand, yes this has to happen on the client PC, since the connection is encrypted. But this fact doesn't count for this tool.

File: flask_login/utils.py (hunk 1 above)
Comment (alanhamlett) [inline flask_login/utils.py]: The login url can also be injected, shown in `test_unauthorized_uses_host_from_next_url`.

**Comment by davidism** [comment]
That appears to be an issue with Werkzeug. It is already reported at pallets/werkzeug#609, with a lot of discussion. Unfortunately, there are not many maintainers and we seem to have lost track of this one.

Flask-Login should not be patching anything here.

Originally, I thought you were referring to the `next` URL, which, can be set to an arbitrary location and is worth looking into.

**Comment by alanhamlett** [comment]
@davidism nice find! I'll close this in favor of werkzeug returning the correct host in `request.url`.

**Pull request closed without merging**